lib/pid/src/PID.cpp

```diff
@@ -1,492 +1,120 @@
-  prevControllerOutput_ *= (outMax - outMin) / outSpan_;
-
-  // Make sure the working variables are within the new limits.
-  if (prevControllerOutput_ > 1) {
-    prevControllerOutput_ = 1;
-  } else if (prevControllerOutput_ < 0) {
-    prevControllerOutput_ = 0;
-  }
-
-  outMin_  = outMin;
-  outMax_  = outMax;
-  outSpan_ = outMax - outMin;
+void PID::PID::updateDerivativeGain(float d) {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_DGain = d;
 }
 
-void PID::setTunings(float Kc, float tauI, float tauD) {
-  // Verify that the tunings make sense.
-  if (Kc == 0.0 || tauI < 0.0 || tauD < 0.0) {
-    return;
-  }
-
-  // Store raw values to hand back to user on request.
-  pParam_ = Kc;
-  iParam_ = tauI;
-  dParam_ = tauD;
-
-  float tempTauR;
-
-  if (tauI == 0.0) {
-    tempTauR = 0.0;
-  } else {
-    tempTauR = (1.0 / tauI) * tSample_.count();
-  }
-
-  // For "bumpless transfer" we need to rescale the accumulated error.
-  if (inAuto) {
-    if (tempTauR == 0.0) {
-      accError_ = 0.0;
-    } else {
-      accError_ *= (Kc_ * tauR_) / (Kc * tempTauR);
-    }
-  }
-
-  Kc_   = Kc;
-  tauR_ = tempTauR;
-  tauD_ = tauD / tSample_.count();
+void PID::PID::updateDeadzone(float deadzone) {
+  std::lock_guard<Mutex> lock(m_mutex);
+  m_deadzone = deadzone;
 }
 
-void PID::reset(void) {
-  float scaledBias = 0.0;
-
-  if (usingFeedForward) {
-    scaledBias = (bias_ - outMin_) / outSpan_;
-  } else {
-    scaledBias = (realOutput_ - outMin_) / outSpan_;
-  }
-
-  prevControllerOutput_ = scaledBias;
-  prevProcessVariable_  = (processVariable_ - inMin_) / inSpan_;
-
-  // Clear any error in the integral.
-  accError_ = 0;
-}
-
-void PID::setMode(int mode) {
-  // We were in manual, and we just got set to auto.
-  // Reset the controller internals.
-  if (mode != 0 && !inAuto) {
-    reset();
-  }
-
-  inAuto = (mode != 0);
+float PID::PID::reportProportionalGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_PGain;
 }
 
-void PID::setInterval(std::chrono::duration<float> interval) {
-  if (interval > 0s) {
-    // Convert the time-based tunings to reflect this change.
-    tauR_ *= (interval / tSample_);
-    accError_ *= (tSample_ / interval);
-    tauD_ *= (interval / tSample_);
-    tSample_ = interval;
-  }
-}
-
-void PID::setSetPoint(float sp) {
-  setPoint_ = sp;
+float PID::PID::reportIntegralGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_IGain;
 }
 
-void PID::setProcessValue(float pv) {
-  processVariable_ = pv;
+float PID::PID::reportDerivativeGain() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_DGain;
 }
 
-void PID::setBias(float bias) {
-  bias_            = bias;
-  usingFeedForward = 1;
+float PID::PID::reportDeadzone() const {
+  std::lock_guard<Mutex> lock(m_mutex);
+  return m_deadzone;
 }
 
-void PID::setDeadZoneError(float error) {
-  deadZoneError_ = error;
+void PID::PID::reset() {
+  std::lock_guard<Mutex> lock(m_mutex);
```

Comment:
bump on the scoped lock comment

Reply:
wdym bump?

Reply:
Look at my previous comment about lockguard vs scoped lock

Reply:
I dont see any comment abt scoped lock